src/check/canonicalize.zig

```diff
@@ -513,411 +486,248 @@ pub fn canonicalize_expr(
                 .list = .{
                     .elems = elems_span,
                     .elem_var = fresh_type_var,
+                    .region = self.tokenizedRegionToRegion(e.region),
                 },
-            };
-
-            return self.can_ir.store.addExpr(.{
-                .expr = list_expr,
-                .region = e.region.toBase(),
             });
         },
         .tag => |e| {
             if (self.parse_ir.tokens.resolveIdentifier(e.token)) |tag_name| {
                 const fresh_type_var_tag_union = self.can_ir.env.types_store.fresh();
                 const fresh_type_var_ext = self.can_ir.env.types_store.fresh();
 
-                const tag_expr = CIR.Expr{
+                return self.can_ir.store.addExpr(CIR.Expr{
                     .tag = .{
                         .tag_union_var = fresh_type_var_tag_union,
                         .ext_var = fresh_type_var_ext,
                         .name = tag_name,
                         .args = .{ .span = .{ .start = 0, .len = 0 } }, // empty arguments
+                        .region = self.tokenizedRegionToRegion(e.region),
                     },
-                };
-
-                return self.can_ir.store.addExpr(.{
-                    .expr = tag_expr,
-                    .region = e.region.toBase(),
                 });
             } else {
                 return null;
             }
         },
-        .string_part => |e| {
-            _ = self.can_ir.env.problems.append(self.can_ir.env.gpa, Problem.Canonicalize.make(.{ .NotYetImplementedExpr = .{
-                .expr_type = "string_part",
-                .region = e.region.toBase(),
-            } }));
+        .string_part => |_| {
+            self.can_ir.env.pushProblem(Problem.Compiler.can(.not_implemented));
             return null;
         },
-        .tuple => |e| {
-            _ = self.can_ir.env.problems.append(self.can_ir.env.gpa, Problem.Canonicalize.make(.{ .NotYetImplementedExpr = .{
-                .expr_type = "tuple",
-                .region = e.region.toBase(),
-            } }));
+        .tuple => |_| {
+            self.can_ir.env.pushProblem(Problem.Compiler.can(.not_implemented));
             return null;
         },
-        .record => |e| {
-            _ = self.can_ir.env.problems.append(self.can_ir.env.gpa, Problem.Canonicalize.make(.{ .NotYetImplementedExpr = .{
-                .expr_type = "record",
-                .region = e.region.toBase(),
-            } }));
+        .record => |_| {
+            self.can_ir.env.pushProblem(Problem.Compiler.can(.not_implemented));
             return null;
         },
         .lambda => |e| {
-            _ = self.can_ir.env.problems.append(self.can_ir.env.gpa, Problem.Canonicalize.make(.{ .NotYetImplementedExpr = .{
-                .expr_type = "lambda",
-                .region = e.region.toBase(),
-            } }));
-            return null;
+
+            // args
```

Comment:
Should there be a TODO below this comment?

Reply:
yeah, I was part way through implementing the analysis for lambdas and realised I wasn't sure if we needed another variant in Expr or if there was something else we needed to do. But yeah, we need to handle the args at this point.

Reply:
resolved in commit d936f9ccc72ad4fdfd76b2debb8147780fae7d2f